lib/src/socket.dart

```
@@ -392,211 +335,217 @@ class PhoenixSocket {
-          error: err,
-          stacktrace: stacktrace,
+      _stateEventStreamController.add(
+        PhoenixSocketErrorEvent(
+          error: error,
+          stacktrace: stackTrace,
         ),
-      ));
-
-      return false;
-    } catch (err, stacktrace) {
-      _logger.severe(
-        '[phoenix_socket] Heartbeat message failed',
-        err,
-        stacktrace,
       );
 
+      return false;
+    } catch (error, stackTrace) {
+      _logger.warning('Heartbeat message failed', error, stackTrace);
+      _reconnect(4001, reason: 'Heartbeat timeout');
       return false;
     }
   }
 
-  void _triggerChannelExceptions(PhoenixException exception) {
-    _logger.fine(
-      () => 'Trigger channel exceptions on ${channels.length} channels',
-    );
-    for (final channel in channels.values) {
-      _logger.finer(
-        () => 'Trigger channel exceptions on ${channel.topic}',
+  Timer _scheduleHeartbeat([String? previousHeartbeat]) {
+    return Timer(_options.heartbeat, () {
+      if (previousHeartbeat != null) {
+        final completer = _pendingMessages.remove(previousHeartbeat);
+        if (completer != null && !completer.isCompleted) {
+          completer.completeError(
+            TimeoutException(
+              'Heartbeat $previousHeartbeat not completed before sending new one',
+            ),
+            StackTrace.current,
+          );
+          return;
+        }
+      }
+
+      _sendHeartbeat();
+    });
+  }
+
+  void _cancelHeartbeat() {
+    if (_heartbeatTimeout != null) {
+      _heartbeatTimeout?.cancel();
+      _heartbeatTimeout = null;
+    }
+    if (_latestHeartbeatRef != null) {
+      _pendingMessages.remove(_latestHeartbeatRef)?.complete(
+            Message.heartbeat(_latestHeartbeatRef!), // doesn't matter
+          );
+      _latestHeartbeatRef = null;
+    }
+  }
+
+  bool _shouldPipeMessage(String message) {
+    final currentSocketState = _socketStateStream.valueOrNull;
```

Oh, forgot to leave a comment here. This could become a situation for a "black hole" where a user isn't receiving events but "we don't know why."

I can't say this will be a problem, but I wonder if we should throw an exception or log something here for the first version to ensure we aren't dropping anything we might want to keep.

---

This is very much in line with how it used to work:
```dart
bool _shouldPipeMessage(dynamic event) {
  if (event is WebSocketChannelException) {
    return true;
  } else if (_socketState != SocketState.closed) {
    return true;
  } else {
    _logger.warning(
      'Message from socket dropped because PhoenixSocket is closed',
      '  $event',
    );
    return false;
  }
}
```

I agree that this is a bit dubious, and maybe we can change the `if` condition to allow also the disconnecting state.